src/link/script.y

```diff
@@ -0,0 +1,479 @@
+%language "c++"
+%define api.value.type variant
+%define api.token.constructor
+
+%code requires {
+	#include <stdint.h>
+	#include <string>
+
+	#include "linkdefs.hpp"
+}
+%code {
+	#include <algorithm>
+	#include <array>
+	#include <assert.h>
+	#include <bit>
+	#include <cinttypes>
+	#include <fstream>
+	#include <locale>
+	#include <string_view>
+	#include <vector>
+
+	#include "helpers.hpp"
+	#include "itertools.hpp"
+	#include "util.hpp"
+
+	#include "link/main.hpp"
+	#include "link/section.hpp"
+
+	using namespace std::literals;
+
+	static void includeFile(std::string &&path);
+	static void incLineNo(void);
+
+	static void setSectionType(SectionType type);
+	static void setSectionType(SectionType type, uint32_t bank);
+	static void setAddr(uint32_t addr);
+	static void alignTo(uint32_t alignment, uint32_t offset);
+	static void pad(uint32_t length);
+	static void placeSection(std::string const &name);
+
+	static yy::parser::symbol_type yylex(void);
+
+	struct Keyword {
+		std::string_view name;
+		yy::parser::symbol_type (* tokenGen)(void);
+	};
+}
+
+%token YYEOF 0
+%token newline
+%token ORG "ORG"
+       INCLUDE "INCLUDE"
+       ALIGN "ALIGN"
+       DS "DS"
+%code {
+	static std::array keywords{
+		Keyword{"ORG"sv,     yy::parser::make_ORG},
+		Keyword{"INCLUDE"sv, yy::parser::make_INCLUDE},
+		Keyword{"ALIGN"sv,   yy::parser::make_ALIGN},
+		Keyword{"DS"sv,      yy::parser::make_DS},
+	};
+}
+%token <std::string> string;
+%token <uint32_t> number;
+%token <SectionType> section_type;
+
+%%
+
+lines: %empty
+     | line lines
+;
+
+line: INCLUDE string newline { includeFile(std::move($2)); } // Note: this additionally increments the line number!
+    | directive newline { incLineNo(); }
```

Comment:
These definitions make it an outright syntax error to leave out the final newline, not even just a "no newline at end of file" warning. it also just prints "`(null)`" for the filename.

```
$ cat foo.asm
section "test", rom0
db $aa, $bb, $cc

$ cat foo.link
ROM0
$ cat foo2.inc
"test"

$ ./rgbasm -o foo.o foo.asm
$ ./rgblink -x -l foo.link -o foo.gb foo.o
error: (null)(289): syntax error, unexpected YYEOF, expecting newline
Linking failed with 1 error
```

Reply:
That `(null)` is actually UB... that's annoying, too. I'll see what I can do, but do you mind making this into a proper test case, in the meantime?

(Note that this will largely echo the work we've had to carry for RGBASM's `INCLUDE`-at-EOF shenanigans... But I'll look for a cleaner solution; hopefully there's one.)

Reply:
Added a test case.

Reply:
Fixed it!